onnx/shape_inference/shape_inference.h

```diff
@@ -77,22 +77,60 @@ void InferShapes(ModelProto& m) {
 
       schema->GetShapeInferenceFunction()(ctx);
 
-      // TODO: this handles the simple case of adding value_infos when
-      // there were none before. It should be enhanced to refine
-      // partially-specified types as well.
-
       for (int i = 0; i < n.output_size(); i++) {
         const auto& output = n.output(i);
-        if (ctx.getOutputType(i)->elem_type() == TensorProto::UNDEFINED) {
+        const auto& inferredType = *ctx.getOutputType(i);
+
+        // In this case, we have no new information, so don't bother
+        // to add a contentless ValueInfo.
+        if (inferredType.elem_type() == TensorProto::UNDEFINED &&
+            inferredType.shape().dim_size() == 0) {
           continue;
         }
 
-        if (valueTypesByName.find(output) == valueTypesByName.end()) {
-          valueTypesByName[output] = *ctx.getOutputType(i);
-          auto* vi = g->add_value_info();
+        // If there is already a ValueInfo associated with this
+        // output, reuse it. Otherwise add a new one.
+        auto iter = valueTypesByName.find(output);
+        TypeProto_Tensor* existingType = nullptr;
+        if (iter != valueTypesByName.end()) {
+          existingType = iter->second;
+        } else {
+          auto vi = g->add_value_info();
           vi->set_name(output);
-          *vi->mutable_type()->mutable_tensor_type() = *ctx.getOutputType(i);
+          existingType = vi->mutable_type()->mutable_tensor_type();
+        }
+
+        // Incorporate the inferred information.
+        if (inferredType.elem_type() != TensorProto::UNDEFINED) {
+          if (existingType->elem_type() != TensorProto::UNDEFINED &&
+              existingType->elem_type() != inferredType.elem_type()) {
+            std::cerr << "Warning: inferred type differs from existing type" << std::endl;
+          } else {
+            existingType->set_elem_type(inferredType.elem_type());
+          }
         }
+        if (inferredType.shape().dim_size() != 0) {
+          if (existingType->shape().dim_size() == 0) {
+            for (int j = 0; j < inferredType.shape().dim_size(); j++) {
+              existingType->mutable_shape()->add_dim();
+            }
+          }
+          for (int j = 0; j < inferredType.shape().dim_size(); j++) {
+            const auto& inferredDim = inferredType.shape().dim(j);
+            auto* existingDim = existingType->mutable_shape()->mutable_dim(j);
+            if (inferredDim.has_dim_value()) {
```

Comment:
you don't handle dim_param here

Reply:
yes, I intentionally left that out for now. If I infer a shape of `(3, x, 10)`, it's not quite clear what to do with the `x`, so I just leave the second dimension alone